adapters/flipp/flipp.go

```diff
@@ -0,0 +1,220 @@
+)
+
+type adapter struct {
+	endpoint string
+}
+
+// Builder builds a new instance of the Flipp adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	bidder := &adapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	adapterRequests := make([]*adapters.RequestData, 0, len(request.Imp))
+	var errors []error
+
+	for _, imp := range request.Imp {
+		adapterReq, err := a.processImp(request, imp)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		adapterRequests = append(adapterRequests, adapterReq)
+	}
+	if len(adapterRequests) == 0 {
+		return nil, append(errors, fmt.Errorf("adapterRequest is empty"))
+	}
+	return adapterRequests, errors
+}
+
+func (a *adapter) makeRequest(request *openrtb2.BidRequest, campaignRequestBody CampaignRequestBody) (*adapters.RequestData, error) {
+	campaignRequestBodyJSON, err := json.Marshal(campaignRequestBody)
+	if err != nil {
+		return nil, err
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json")
+	headers.Add("User-Agent", request.Device.UA)
+	return &adapters.RequestData{
+		Method:  http.MethodPost,
+		Uri:     a.endpoint,
+		Body:    campaignRequestBodyJSON,
+		Headers: headers,
+	}, err
+}
+
+func (a *adapter) processImp(request *openrtb2.BidRequest, imp openrtb2.Imp) (*adapters.RequestData, error) {
+	var flippExtParams openrtb_ext.ImpExtFlipp
+	params, _, _, err := jsonparser.Get(imp.Ext, "bidder")
+	if err != nil {
+		return nil, fmt.Errorf("flipp params not found. %v", err)
+	}
+	err = json.Unmarshal(params, &flippExtParams)
+	if err != nil {
+		return nil, fmt.Errorf("unable to extract flipp params. %v", err)
+	}
+
+	publisherUrl, err := url.Parse(request.Site.Page)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse site url. %v", err)
+	}
+
+	var contentCode string
+	if publisherUrl != nil {
+		contentCode = publisherUrl.Query().Get("flipp-content-code")
+	}
+
+	placement := Placement{
+		DivName: inlineDivName,
+		SiteID:  &flippExtParams.SiteID,
+		AdTypes: getAdTypes(flippExtParams.CreativeType),
+		ZoneIds: flippExtParams.ZoneIds,
+		Count:   &count,
+		Prebid:  buildPrebidRequest(flippExtParams, request, imp),
+		Properties: &Properties{
+			ContentCode: &contentCode,
+		},
+		Options: flippExtParams.Options,
+	}
+
+	var userKey string
+	if request.User != nil && request.User.ID != "" {
+		userKey = request.User.ID
+	} else if flippExtParams.UserKey != "" {
+		userKey = flippExtParams.UserKey
+	} else {
+		uid, err := uuid.NewV4()
+		if err != nil {
+			return nil, fmt.Errorf("unable to generate user uuid. %v", err)
+		}
+		userKey = uid.String()
+	}
+
+	keywordsArray := strings.Split(request.Site.Keywords, ",")
+	var keywordsSlice []Keyword
```

Comment:
Can you please help me understand  why are we using a custom data type and not using string directly?

Reply:
@Sonali-More-Xandr thanks just fixed this